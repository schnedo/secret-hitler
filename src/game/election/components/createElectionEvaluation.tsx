import type { ComponentType, ReactElement } from "react";
import type { ModalProps } from "../../../components";
import type PlayerVotes from "../../PlayerVotes";
import type Phase from "../../Phase";
import type { Player } from "../../player";

export interface ElectionEvaluationProps {
  onElectionAccepted: () => void;
  onElectionDeclined: () => void;
  phase: Phase;
  playerVotes: PlayerVotes;
  players: Player[];
}

export default function createElectionEvaluation(
  Modal: ComponentType<ModalProps>,
) {
  return function ElectionEvaluation({
    onElectionAccepted,
    onElectionDeclined,
    phase,
    playerVotes,
    players,
  }: ElectionEvaluationProps): ReactElement {
    if (phase !== "electionEvaluation") {
      return <></>;
    }
    const electionAccepted =
      Object.values(playerVotes).filter(Boolean).length > players.length / 2;
    const continueAction = electionAccepted
      ? onElectionAccepted
      : onElectionDeclined;

    return (
      <Modal open>
        <div>Election {electionAccepted ? "" : "not"} accepted</div>
        {Object.entries(playerVotes).map(([playerId, accepted]) => (
          <div key={playerId}>
            {players[parseInt(playerId)].name}: {accepted ? "Yes" : "No"}
          </div>
        ))}
        <button onClick={continueAction}>continue</button>
      </Modal>
    );
  };
}
