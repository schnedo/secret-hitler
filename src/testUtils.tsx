import type { PropsWithChildren } from "react";

export function MockComponent({
  children,
  ...props
}: PropsWithChildren<unknown>): JSX.Element {
  return (
    <div>
      {JSON.stringify(props)} {children}
    </div>
  );
}
