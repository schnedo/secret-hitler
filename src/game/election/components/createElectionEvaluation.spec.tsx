import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { MockComponent } from "../../../testUtils";
import type { Player } from "../../player";
import type PlayerVotes from "../PlayerVotes";
import createElectionEvaluation from "./createElectionEvaluation";

const ElectionEvaluation = createElectionEvaluation(MockComponent);

const somePlayers: Player[] = [
  {
    name: "John",
    role: "fascist",
    title: null,
  },
  {
    name: "Martha",
    role: "hitler",
    title: null,
  },
  {
    name: "Bob",
    role: "liberal",
    title: null,
  },
  {
    name: "Alice",
    role: "liberal",
    title: null,
  },
  {
    name: "Mohammed",
    role: "liberal",
    title: null,
  },
];
const allYesVotes: PlayerVotes = somePlayers.map(() => true);

it("should render correctly", async () => {
  expect.hasAssertions();

  const { container } = render(
    <ElectionEvaluation
      onElectionAccepted={() => undefined}
      onElectionDeclined={() => undefined}
      playerVotes={allYesVotes}
      players={somePlayers}
    />,
  );

  expect(container).toMatchSnapshot();
});

it("should call onElectionAccepted when election got accepted", async () => {
  expect.hasAssertions();

  const handleElectionAccepted = jest.fn();
  render(
    <ElectionEvaluation
      onElectionAccepted={handleElectionAccepted}
      onElectionDeclined={() => undefined}
      playerVotes={allYesVotes}
      players={somePlayers}
    />,
  );

  expect(handleElectionAccepted).not.toHaveBeenCalled();
  await userEvent.click(screen.getByRole("button"));
  expect(handleElectionAccepted).toHaveBeenCalledTimes(1);
});

it("should call onElectionDeclined when election got accepted", async () => {
  expect.hasAssertions();

  const allNoVotes = somePlayers.map(() => false);
  const handleElectionDeclined = jest.fn();
  render(
    <ElectionEvaluation
      onElectionAccepted={() => undefined}
      onElectionDeclined={handleElectionDeclined}
      playerVotes={allNoVotes}
      players={somePlayers}
    />,
  );

  expect(handleElectionDeclined).not.toHaveBeenCalled();
  await userEvent.click(screen.getByRole("button"));
  expect(handleElectionDeclined).toHaveBeenCalledTimes(1);
});
