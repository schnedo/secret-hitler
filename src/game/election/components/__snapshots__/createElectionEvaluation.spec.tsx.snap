// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div>
  <div>
    {"open":true}
     
    <div>
      Election 
      
       accepted
    </div>
    <div>
      John
      : 
      Yes
    </div>
    <div>
      Martha
      : 
      Yes
    </div>
    <div>
      Bob
      : 
      Yes
    </div>
    <div>
      Alice
      : 
      Yes
    </div>
    <div>
      Mohammed
      : 
      Yes
    </div>
    <button>
      continue
    </button>
  </div>
</div>
`;
